### PR TITLE
#2171: Implement addUserDefinedData function to PhaseManager

### LINKED ISSUE #2171: Add application-specific information to phase (e.g., timesteps) in JSON output
**What Needs to be Done?**

We want to be able to match up timesteps in NimbleSM with phases IDs.

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #2171

### CODE REVIEW

**Review by lifflander — APPROVED** [review]
Looks good to me

File: src/vt/phase/phase_manager.h [hunk 1]
```diff
@@ -169,6 +170,30 @@ struct PhaseManager : runtime::component::Component<PhaseManager> {
...
+  );
+  static_assert(
+    is_jsonable<ValueT>::value,
+    "PhaseManager::addUserDefinedData: ValueT type is not jsonable"
+  );
+
+  auto j = std::make_shared<nlohmann::json>();
+  j->emplace(key, value);
+
+  theNodeLBData()->getLBData()->user_per_phase_json_[phase] = j;
```
Comment (nlslatt) [inline src/vt/phase/phase_manager.h:194]: Does this approach limit the user-defined data to a single key-value pair? I think we should allow as much as the application wants to provide.

**Review by nlslatt — CHANGES_REQUESTED** [review]
Please see concern below.

File: src/vt/phase/phase_manager.h (hunk 1 above)
Comment (JacobDomagala) [reply src/vt/phase/phase_manager.h]: Updated

**Review by nlslatt — APPROVED** [review]
Looks good

**Pull request merged**